src/sentry/services/hybrid_cloud/user/impl.py

```diff
@@ -181,18 +181,16 @@ def serialize_rpc_user(user: User) -> RpcUser:
         roles = frozenset(flatten(user.roles))
     args["roles"] = roles
 
-    useremails: FrozenSet[RpcUserEmail] = frozenset({})
+    useremails: List[RpcUserEmail] = []
```

Comment:
I'm curious why this change was necessary?  Because the actual model does require `frozensets` due to the requirement of `hash` for these models.

Reply:
Pydantic models, like dataclasses without `(frozen=True, eq=True)`, aren't hashable by default. I get that these `frozenset`s are meant to make `RpcUser` hashable, but they also rely on `RpcUserEmail` being hashable, which isn't either.

They *can* be made hashable (simply by defining a custom `__hash__` method) but I'm experimenting with making it unnecessary. Seems better to go with Pydantic's flow, in that regard. If we actually need logical sets (i.e., discarding duplicate instances of equal objects), it might be simpler and clearer in those cases to explicitly build a dict indexed by ID.

On a philosophical note, I generally like hashable classes better than non-hashable ones, but I also want my hashable classes to be immutable. I vaguely remember a conversation a while back about the advantages of making these models mutable after initialization. Is that still the case?

Reply:
Absolutely.  IT's kind of a mess.  My ideal world was that the Rpc models had immutable base states and mutable builders, to match the reality of how these classes get used.  That was our older conversation

Part of the trick is that our Rpc classes are often fitting into existing codes and "duck typing" some other type, that may or may not be used in various ways (unfortunately, grabbing a set reference and mutating it alongside doing some work isn't out of the question, or using these classes with hashs as keys or set entries, expecting identifier deduping).  We've often made various decisions to try to imitate the existing behaviors and interfaces of those classes to reduce the impact of our "two code paths" approach (existing SaaS vs incremental region).  So it's been tactical.

I believe that once the migration to 100% regional silo logic is complete, we can untangle this mess by rewriting different parts of the code to remove this.  In the mean time, we have to pick our fights.

That said, I don't have a strong opinion other than we have to be mindful of how these classes are used and find an appropriate solution.  If a custom hash is the best option, I can be convinced of that.

Reply:
tldr: ok, that's fine by me.  I just want to make sure you're aware and we do something to address the existing usage.  There should, be test coverage.  And you may be seeing those tests fail right now.

Reply:
Also, as far as "where" models are used as keys...

the answer is:
1.  In notifications code, all over the place and,
2.  In virtually every serializer code path 🙄

Reply:
Yep, rest assured I'm doing it consciously. I checked `serialize_rpc_user` and there seems to be no risk of a behavioral change there. Not sure about where we care about `RpcUser` being hashed exactly, but I'll follow up if I don't run across it in CI.